psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1600 @@
+					integrals[ (Shell_M_start + func_m) * nbf_ + Shell_N_start + func_n ]
+					=
+//					integrals[ (Shell_N_start + func_n) * nbf_ + Shell_M_start + func_m ]
+//					=
+					val 
+					= 
+ fabs(buffer[rank][ func_m * Shell_N_count * Shell_M_count * Shell_N_count + 
+				    func_n * Shell_M_count * Shell_N_count + 
+				    func_m * Shell_N_count + 
+				    func_n]);
+					if (global_max_int < val) global_max_int = val;
+					
+					if (shel_maxes[Shell_M * p_shells_ + Shell_N] < val) {
+						shel_maxes[Shell_M * p_shells_ + Shell_N]
+//						=
+//						shel_maxes[Shell_N * p_shells_ + Shell_M]
+						= val;
+					}
+				}
+			}
+		}
+	}
+
+	tolerance_ = cutoff_ * cutoff_ / global_max_int;
+	schwarz_shell_mask_pQq_.resize(p_shells_);
+	schwarz_func_starts_pQq_.resize(p_shells_);
+	schwarz_dense_funcs_.resize(p_shells_);
+	size_t count;
+
+	for ( size_t shell_it_out = 0U; shell_it_out < p_shells_; shell_it_out++ ) {
+		count = 0U;
+		for (size_t shell_it_in = 0U; shell_it_in < p_shells_; shell_it_in++ ) {
+			if(shel_maxes[shell_it_out*p_shells_ + shell_it_in] >  tolerance_) {
+				schwarz_shell_mask_pQq_[shell_it_out].push_back(shell_it_in);
+				schwarz_func_starts_pQq_[shell_it_out].push_back(count);
+				count += primary_->shell(shell_it_in).nfunction();
+			}
+		}
+		schwarz_dense_funcs_[shell_it_out] = count;
+	}
+}
+
+// calculates a block of AO's then conducts the proper contractions to get them
+//   where they need to go. I don't think it makes sense to have seperate J and
+//   K function calls as J and K are never fully built in a single function.
+void DirectDFJK::compute_JK() {
+	BB_ = false;
+
+	if ( pQq_ ) {
+        if (ao_sparse_) {
+		    pQp_sparse();
+        } else {
+    	    pQp();
+        }
+
+	}  else {
+
+		if (num_blocks_ == 1) {
+			build_jk_CC_Qpq_direct();
+		} else {
+    		build_jk_CC_Qpq_blocks();
+		}
+
+	}
+    n_col_last_ = C_left_ao_[0]->ncol();
+}
+
+void DirectDFJK::postiterations() {
+}
+
+// Determines the maximum amount of memory that DirectDFJK will be using
+//   at any one time. To be sure, the memory used to construct AO blocks
+//   will only be allocated during the compute_JK() method, so it will
+//   be available during other procedures. 
+// Uses blocking and sparsity information to set total_needs which is 
+//   returned by memory_estimate.
+void DirectDFJK::our_needs(){
+	free_memory_ = memory_;
+	size_t blocks;
+
+	size_t our_needs = 0;
+	double charges_f = 0.0;
+	size_t charges_z;
+	std::shared_ptr<Molecule> mol_ptr = primary_->molecule();
+	
+	for (int atom_iter = 0; atom_iter < mol_ptr->natom(); atom_iter++)
+		charges_f+=mol_ptr->fZ(atom_iter);
+
+	charges_f = charges_f/2.0;
+
+	charges_z = static_cast<size_t>(charges_f);
+	++charges_z;
+
+	//D
+	our_needs += nbf_*nbf_;
+	free_memory_-=nbf_*nbf_;
+
+	//C_a
+	our_needs += nbf_*charges_z;
+	free_memory_-=nbf_*charges_z;
+	//C_b
+	our_needs += nbf_*charges_z;
+	free_memory_-=nbf_*charges_z;
```

Comment:
So... you're assuming a neutral molecule and you can fit all half-transformed integrals in memory? Both are no-no's.

Reply:
Integrals are not stored between SCF iterations. This is only meant as a rough estimate so that the DFT grid doesn't take all the memory. The code in the scf iterations allocates space based on the size of the actual SCF coefficient matrix from iteration to iteration, so this isn't a problem.